File: swift/Sources/UserIntentKit/WatchAdapter.swift
// SPDX-License-Identifier: AGPL-3.0

import Foundation

/// Watch adapter for Apple Watch / Wear OS.
/// Publishes watch-specific signals (wrist raise, screen active, context)
/// and reads derived state for alert routing.
public class WatchAdapter {
    private let client: IntentClient
    private var publishTimer: Timer?

    public init(client: IntentClient) {
        self.client = client
    }

    /// Publish current watch state to intent API.
    public func publishState(
        context: String = "active",
        screenActive: Bool = true,
        wristRaise: Bool = false,
        custom: [String: Any]? = nil
    ) async throws {
        var fields: [String: Any] = [
            "context": context,
            "screen_active": screenActive,
            "wrist_raise": wristRaise,
        ]
        if let custom {
            fields["custom"] = custom
        }
        try await client.patchDevice(fields)
    }

    /// Check how an urgent alert should be delivered.
    public func alertMode() async throws -> AlertMode {
        let derived = try await client.getDerived()
        let urgency = derived["urgency_mode"] as? String ?? "normal"
        let suppressAudio = derived["suppress_audio"] as? Bool ?? false

        switch urgency {
        case "emergency-only":
            return .silent
        case "text-only":
            return .textOnly
        default:
            return suppressAudio ? .textOnly : .full
        }
    }

    /// Start periodic state publishing and heartbeat.
    public func start(interval: TimeInterval = 30.0) {
        stop()
        client.startHeartbeat(interval: interval)
        publishTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task {
                try? await self?.publishState()
            }
        }
    }

    public func stop() {
        publishTimer?.invalidate()
        publishTimer = nil
        client.stopHeartbeat()
    }
}

public enum AlertMode {
    /// Full alert: vibrate + text + optional audio
    case full
    /// Text only: vibrate + text card, no audio
    case textOnly
    /// Silent: no alert unless emergency
    case silent
}

File: kotlin/src/main/kotlin/com/thinkoff/userintent/WatchAdapter.kt
// SPDX-License-Identifier: AGPL-3.0

package com.thinkoff.userintent

import kotlinx.coroutines.*

/**
 * Watch adapter for Wear OS smartwatches.
 * Publishes watch signals and reads alert mode for notification routing.
 */
class WatchAdapter(
    private val client: IntentClient,
    private val scope: CoroutineScope
) {
    private var publishJob: Job? = null

    /**
     * Publish current watch state to intent API.
     */
    suspend fun publishState(
        context: String = "active",
        screenActive: Boolean = true,
        wristRaise: Boolean = false,
        custom: Map<String, Any?>? = null
    ) {
        val fields = mutableMapOf<String, Any?>(
            "context" to context,
            "screen_active" to screenActive,
            "wrist_raise" to wristRaise
        )
        if (custom != null) fields["custom"] = custom
        client.patchDevice(fields)
    }

    /**
     * Check how an urgent alert should be delivered.
     */
    suspend fun alertMode(): AlertMode {
        val derived = client.getDerived()
        val urgency = derived["urgency_mode"] as? String ?: "normal"
        val suppressAudio = derived["suppress_audio"] == true

        return when {
            urgency == "emergency-only" -> AlertMode.SILENT
            urgency == "text-only" || suppressAudio -> AlertMode.TEXT_ONLY
            else -> AlertMode.FULL
        }
    }

    /**
     * Start periodic state publishing and heartbeat.
     */
    fun start(intervalMs: Long = 30000L) {
        stop()
        client.startHeartbeat(scope)
        publishJob = scope.launch {
            while (isActive) {
                try { publishState() } catch (_: Exception) {}
                delay(intervalMs)
            }
        }
    }

    fun stop() {
        publishJob?.cancel()
        publishJob = null
        client.stopHeartbeat()
    }
}

enum class AlertMode {
    /** Full alert: vibrate + text + optional audio */
    FULL,
    /** Text only: vibrate + text card, no audio */
    TEXT_ONLY,
    /** Silent: no alert unless emergency */
    SILENT
}
